Assets/Scripts/ProceduralGeneration.cs

```diff
@@ -0,0 +1,192 @@
+        bounds = new BoundingSphere[maxEntities];
+        cullGroup.SetBoundingSpheres(bounds);
+        cullGroup.SetBoundingSphereCount(0);
+        cullGroup.onStateChanged += CullingStateChanged;
+        entityLookup = new Dictionary<int, GameObject>();
+        entityTransforms = new Dictionary<int, Transform>();
+        entityIdToIndex = new Dictionary<int, int>();
+        entityIndexToId = new Dictionary<int, int>();
+
+        StartCoroutine(nameof(GenerateAsteroidField));
+        StartCoroutine(nameof(UpdateCulledObjectBounds));
+    }
+
+    public void AddEntity(GameObject obj)
+    {
+        var id = obj.GetInstanceID();
+        var tf = obj.transform;
+        var radius = obj.GetComponent<Renderer>().bounds.size.x;
+        entityLookup.Add(id, obj);
+        var index = EntityCount - 1;
+        bounds[index] = new BoundingSphere {position = tf.position, radius = radius};
+        entityTransforms.Add(id, tf);
+        entityIndexToId.Add(index, id);
+        entityIdToIndex.Add(id, index);
+        cullGroup.SetBoundingSphereCount(EntityCount);
+    }
+
+    public void RemoveEntity(GameObject obj)
+    {
+        var id = obj.GetInstanceID();
+        var index = entityIdToIndex[id];
+        entityIdToIndex.Remove(id);
+        entityIndexToId.Remove(index);
+        entityTransforms.Remove(id);
+        
+        var moveIndex = EntityCount - 1;
+        if (index != moveIndex)
+        {
+            bounds[index] = bounds[moveIndex];
+            var moveId = entityIndexToId[moveIndex];
+            entityIdToIndex[moveId] = index;
+            entityIndexToId.Remove(moveIndex);
+            entityIndexToId.Add(index, moveId);
+        }
+        entityLookup.Remove(id);
+        cullGroup.SetBoundingSphereCount(EntityCount);
+    }
+
+    private IEnumerator GenerateAsteroidField()
+    {
+        for (var x = 0; x < numChunksInLevel.x; x++)
+        {
+            for (var y = 0; y < numChunksInLevel.y; y++)
+            {
+                for (var z = 0; z < numChunksInLevel.z; z++)
+                {
+                    var chunkStart = new Vector3Int(x, y, z) * chunkSize;
+                    GenerateChunk(chunkStart);
+                    yield return null;
+                }
+            }
+        }
+        FinishedGenerating = true;
+        Debug.Log("Finished Generating!!!");
+    }
+
+    private void GenerateChunk(Vector3Int chunkStart)
+    {
+        for (var x = 0; x < numUnitCubesInChunk.x; x++)
+        {
+            for (var y = 0; y < numUnitCubesInChunk.y; y++)
+            {
+                for (var z = 0; z < numUnitCubesInChunk.z; z++)
+                {
+                    var unitCubeOffset = chunkStart - Vector3.one * (chunkSize / 2 - unitCubeSize / 2);
+                    var unitCubeCenter = (new Vector3Int(x, y, z) * unitCubeSize) + unitCubeOffset;
+                    GenerateChunkAsteroids(unitCubeCenter);
```

Comment:
If we need to, we could extend the algorithm to basically choose between a few different generation functions:
ie GenerateAliens, GenerateMedicalCanisters, GenerateWhatever